boa/integrations/jupyter/jupyter.js

```diff
@@ -3,62 +3,98 @@
     /** Calls the callback endpoint with the given token and body */
     async function callbackAPI(token, body) {
         const headers = {['X-XSRFToken']: getCookie('_xsrf')};
-        const init = {method: 'POST', body: stringify(body), headers};
+        const init = {method: 'POST', body, headers};
         const url = `../titanoboa_jupyterlab/callback/${token}`;
         const response = await fetch(url, init);
         return response.text();
     }
 
     /** Load the signer via ethers user */
     const loadSigner = async () => {
-        const provider = getEthersProvider();
-        console.log(`Loading the user's signer`);
-        const signer = await provider.getSigner();
+        const signer = await getEthersProvider().getSigner();
         return signer.getAddress();
     };
 
     /** Sign a transaction via ethers */
     async function signTransaction(transaction) {
-        const provider = getEthersProvider();
-        console.log('Starting to sign transaction');
-        const signer = await provider.getSigner();
+        const signer = await getEthersProvider().getSigner();
         return signer.sendTransaction(transaction);
     }
 
+    /** Call an RPC method via ethers */
+    const rpc = (method, params) => getEthersProvider().send(method, params);
+
+    /** Wait until the transaction is mined */
+    const waitForTransactionReceipt = async (params, timeout, poll_latency) => {
+        while (true) {
+            try {
+                const result = await rpc('eth_getTransactionReceipt', params);
+                if (result) {
+                    return result;
+                }
+            } catch (err) { // ignore "server error" (happens while transaction is mined)
+                if (err?.info?.error?.code !== -32603) {
+                    throw err;
+                }
+            }
+            if (timeout < poll_latency) {
+                throw new Error('Timeout waiting for transaction receipt');
+            }
+            await sleep(poll_latency);
+            timeout -= poll_latency;
+        }
+    };
+
+    /** Call multiple RPCs in sequence */
+    const multiRpc = (payloads) => payloads.reduce(
+        async (previousPromise, [method, params]) => [...await previousPromise, await rpc(method, params)],
+        [],
+    );
+
     /** Call the backend when the given function is called, handling errors */
     const handleCallback = func => async (token, ...args) => {
-        const body = await parsePromise(func(...args));
-        if (colab) {
-            // Colab expects the response to be JSON
-            return JSON.stringify(body);
-        }
-        const responseText = await callbackAPI(token, body);
-        console.log(`Callback ${token} => ${responseText}`);
+        const body = stringify(await parsePromise(func(...args)));
+        // console.log(`Boa: ${func.name}(${args.map(a => JSON.stringify(a)).join(',')}) = ${body};`);
+        return colab ? body : callbackAPI(token, body);
```

Comment:
i think we shouldn't return anything, since the return values of the two branches mean different things